### PR TITLE
Introduce Vue Testing Library

### PR DESCRIPTION
**Pull request by EshaanAgg**
## Summary
- Introduced the `Vue Testing Library` package into the codebase for testing
- Add added the supplementary `Jest DOM` and `eslint-jest-dom` packages for writing more expressive tests and maintaining the code quality
- Added tests for three files in the repository
- Added `100%` coverage for the files `AttemptIconDiff` and `AttemptTestDiff` and demonstrated how we could mock the Vuex Router and store them in the same
- Added a simple test suite for the `TriesOverview` component and demonstrated the mocking of `mixins` and other dependency imports in the main component

## References
[Primary proposal and discussion link](https://github.com/learningequality/kolibri-design-system/issues/510)

## Reviewer guidance
cc: @MisRob 

## Testing checklist

- [X] Contributor has fully tested the PR manually
- [X] If there are any front-end changes, before/after screenshots are included
- [X] Critical user journeys are covered by Gherkin stories
- [X] Critical and brittle code paths are covered by unit tests

## PR process

- [ ] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

**Comment by EshaanAgg** [comment]
Let me know how you feel about this! If you would like to see how the `VTL` tests would look for any other use-cases or user journeys, please let me know, and I would be happy to add them to this PR for review! Thanks :)

File: kolibri/core/assets/src/views/ExamReport/__tests__/AttemptIconDiff.spec.js
Comment (AlexVelezLl) [inline kolibri/core/assets/src/views/ExamReport/__tests__/AttemptIconDiff.spec.js]: Is there any particular reason you are prefering the render methods insead of the `screen` instance? Using `screen` is the recommended way for queries, and they use it in the examples they provide in docs: https://testing-library.com/docs/queries/about#using-queries.

The `testing-library` author [also wrote on his blog](https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#not-using-screen): (this blog was about react-testing-library but I think it is the same principle)
> The benefit of using screen is you no longer need to keep the render call destructure up-to-date as you add/remove the queries you need. You only need to type screen. and let your editor's magic autocomplete take care of the rest.
The only exception to this is if you're setting the container or baseElement which you probably should avoid doing (I honestly can't think of a legitimate use case for those options anymore and they only exist for historical reasons at this point).

File: kolibri/core/assets/src/views/ExamReport/__tests__/AttemptIconDiff.spec.js
Comment (AlexVelezLl) [inline kolibri/core/assets/src/views/ExamReport/__tests__/AttemptIconDiff.spec.js]: We also should try to avoid over-using the `queryByTestId` as it is not what the user sees, and we would not be using the advantages of testing-library.

Here is something about the [queries priorities](https://testing-library.com/docs/queries/about/#priority) and which one should we use.
>getByTestId: The user cannot see (or hear) these, so this is only recommended for cases where you can't match by role or text or it doesn't make sense (e.g. the text is dynamic).

In this case I dont think there is a better option as we are just checking if the icon is rendered or not (we could use the `presentation` role that KIcon use but I think that doing a query by that would not make sense in the test), but just to keep in mind :smiley:

**Review by AlexVelezLl — COMMENTED** [review]
Hi! @EshaanAgg the tests look cool :smile:, I just noticed a couple of things, but the rest seems fine to me

File: kolibri/core/assets/src/views/ExamReport/__tests__/AttemptIconDiff.spec.js
Comment (EshaanAgg) [reply kolibri/core/assets/src/views/ExamReport/__tests__/AttemptIconDiff.spec.js]: Not really! It is a force of habit, I guess at this point. But since we are just introducing the library to the codebase, it would obviously be nice if we could set some definite style guidelines. Also the argument about less keystrokes and more autocomplete does make sense to me! I will update the tests to reflect the same.

File: kolibri/core/assets/src/views/ExamReport/__tests__/AttemptIconDiff.spec.js
Comment (EshaanAgg) [reply kolibri/core/assets/src/views/ExamReport/__tests__/AttemptIconDiff.spec.js]: I agree. I tried to find the most accessible role for the icon, but since when the Icon is rendered we get some `svg` content, I decided it's the best to use a test id. in general, I first use the `screen.logTestingPlaygroundURL()` method provided by the Testing library to get the most accessible query selector for the element, and use the same as much as possible.

**Comment by MisRob** [comment]
Thanks a lot @EshaanAgg, I will definitely find time to present this to more teammates in the upcoming week or two. Meanwhile, I will turn this PR into draft, just to indicate it's something we're considering.

**Comment by MisRob** [comment]
Hi @EshaanAgg, we're currently discussing our long-term testing plans and looking at the Testing Library within the context of the current landscape of various tools for various test layers. I just shared your work with all developers on that opportunity. We're currently approaching an important release so it may take some time, but I will definitely follow-up. Thanks a lot again!

File: kolibri/core/package.json
Comment (rtibbles) [inline kolibri/core/package.json]: These should be added as dependencies to the kolibri-tools package, rather than being added here.

Comment (EshaanAgg) [reply kolibri/core/package.json]: Got it. Will update. Thanks :)

**Comment by MisRob** [comment]
Hi @EshaanAgg, we talked about this and we'd love to start using the Testing Library. I will convert this from draft to a regular PR and we will finish review.

**Comment by MisRob** [comment]
@AlexVelezLl I'll take some time to read through it too, however as you have experience with the Testing Library, I'd like to leave the final approval to you. Are all your concerns addressed or is there something else you'd like to provide feedback on now that we know we'd like to merge it?

File: kolibri/core/assets/src/views/ExamReport/__tests__/AttemptIconDiff.spec.js
Comment (AlexVelezLl) [inline kolibri/core/assets/src/views/ExamReport/__tests__/AttemptIconDiff.spec.js]: In te future we can change theme colors, so it is better to use the themeTokens variables, you can import it like this:

```js
import { themeTokens } from 'kolibri-design-system/lib/styles/theme';

const successThemeColor = themeTokens().correct;
```

File: packages/kolibri-tools/package.json
Comment (MisRob) [inline packages/kolibri-tools/package.json:1]: @rtibbles I think the dependencies are fine now, right?

**Review by AlexVelezLl — CHANGES_REQUESTED** [review]
Code looks good to me! Just a little final consideration and we're ready to go!

**Comment by AlexVelezLl** [comment]
@MisRob for the future I think it would be good to set guidelines for testing with vue-testing-library, such as preferring to use the screen object, preferring to use the userEvent library, not overusing tests-ids, etc.

File: kolibri/core/assets/src/views/ExamReport/__tests__/AttemptIconDiff.spec.js
Comment (EshaanAgg) [reply kolibri/core/assets/src/views/ExamReport/__tests__/AttemptIconDiff.spec.js]: Ahh, nice catch. Will update!

File: packages/kolibri-tools/package.json
Comment (rtibbles) [inline packages/kolibri-tools/package.json]: Hrm, I think `rewire` is misleading here - all of the dependencies in this devDependencies section should actually be in the `dependencies` section - kolibri-tools is a library for building, testing etc. so all its testing dependencies are actually dependencies and not just for development.

File: packages/kolibri-tools/package.json
Comment (rtibbles) [reply packages/kolibri-tools/package.json:1]: Not quite - the dependencies should not be in devDependencies.

Comment (EshaanAgg) [reply packages/kolibri-tools/package.json:1]: Got it. Will update.

File: packages/kolibri-tools/package.json
Comment (EshaanAgg) [reply packages/kolibri-tools/package.json]: Got it. I have made the required changes and even migrated the `rewire` package to the dependencies section.

**Comment by MisRob** [comment]
> @MisRob for the future I think it would be good to set guidelines for testing with vue-testing-library, such as preferring to use the screen object, preferring to use the userEvent library, not overusing tests-ids, etc.

@AlexVelezLl Agreed. We have a document in Notion with some testing guidelines I wrote years ago. It could be rewritten or even deleted and replaced completely. I think that most of it won't be relevant anymore since it was primarily concerned with issues that Testing Library resolves by its APIs not letting you reach into internals. I'm not sure how often we'd use it for reference though. Perhaps some live demos would work a bit better. Let's chat about in in more detail on Slack.

**Review by MisRob — APPROVED** [review]
Thanks @EshaanAgg, @AlexVelezLl, and @rtibbles.

I just read through the whole suite. Seeing store, router, and other mocks, I feel confident about this being applicable to all our use-cases. I really enjoyed reading through the test suites, it's all very clear.

I will let clicking the merge button on @rtibbles after confirming that dependencies are now fine.

**Comment by EshaanAgg** [comment]
> > @MisRob for the future I think it would be good to set guidelines for testing with vue-testing-library, such as preferring to use the screen object, preferring to use the userEvent library, not overusing tests-ids, etc.
> 
> @AlexVelezLl Agreed. We have a document in Notion with some testing guidelines I wrote years ago. It could be rewritten or even deleted and replaced completely. I think that most of it won't be relevant anymore since it was primarily concerned with issues that Testing Library resolves by its APIs not letting you reach into internals. I'm not sure how often we'd use it for reference though. Perhaps some live demos would work a bit better. Let's chat about in in more detail on Slack.

Sounds great! Even I just joined the developed Slack today itself, so if there is anything that I can help with about this, please let me know! Thanks for all your time and reviews!

**Review by AlexVelezLl — APPROVED** [review]
lgtm!

**Comment by EshaanAgg** [comment]
Hi! Are there any blockers for this PR? I would love to continue the work on this front and add tests for some other components after the same is merged! 👾

**Comment by MisRob** [comment]
Hi @EshaanAgg, no I was just waiting for @rtibbles to confirm that dependencies are now okay, and we're busy with the new release :)

**Comment by rtibbles** [comment]
Yes, dependencies all seem to be in the right place now.

**Comment by MisRob** [comment]
Good, thanks all! Merging.

File: kolibri/core/assets/src/views/ExamReport/__tests__/TriesOverview.spec.js
Comment (bjester) [inline kolibri/core/assets/src/views/ExamReport/__tests__/TriesOverview.spec.js:29]: Double defaulting here in addition to line 23.

Comment (EshaanAgg) [reply kolibri/core/assets/src/views/ExamReport/__tests__/TriesOverview.spec.js:29]: Nice catch! I will fix it in the PR  #11906. Thanks.

**Pull request merged**
